File: src/ui/controls.ts
import type { ExportMode } from '../board/export';
import type { BoardStats } from '../board/board';
import { PALETTE, STROKE } from '../config';
import { luminance } from '../render/color';
import type { GestureMode, Tool } from '../types';
import { el, paintRange } from './dom';
import { loadSettings, saveSettings, type Settings } from './settings';

export interface ControlHandlers {
  onSettingsChange(settings: Settings, key: keyof Settings): void;
  onUndo(): void;
  onRedo(): void;
  onClear(): void;
  onSave(mode: ExportMode): void;
  onStart(): void;
}

export type GateState = 'idle' | 'busy' | 'error' | 'hidden';

export interface GateContent {
  title: string;
  body: string;
  action?: string;
  fine?: string;
}

const GESTURE_NOTES: Record<GestureMode, string> = {
  pinch: 'Touch your thumb to your index fingertip to ink. Separate them to lift the pen.',
  point: 'Extend your index finger and fold the other three to ink.',
  grip: 'Pinch with your ring and little fingers tucked in. Hardest to trigger by accident.',
  any: 'Any of the three gestures inks. Convenient, but easier to start a stroke by mistake.',
};

/** Wires the dock, the settings panel, the gate and the keyboard to one settings object. */
export class Controls {
  readonly settings: Settings;

  private readonly stage = el('stage');
  private readonly toolButtons = [...el('tools').querySelectorAll<HTMLButtonElement>('.tool')];
  private readonly swatchRow = el('swatches');
  private readonly sizeInput = el<HTMLInputElement>('size');
  private readonly sizeValue = el<HTMLOutputElement>('size-value');
  private readonly undoButton = el<HTMLButtonElement>('undo');
  private readonly redoButton = el<HTMLButtonElement>('redo');
  private readonly clearButton = el<HTMLButtonElement>('clear');
  private readonly saveButton = el<HTMLButtonElement>('save');
  private readonly saveMenuToggle = el<HTMLButtonElement>('save-menu-toggle');
  private readonly saveMenu = el('save-menu');

  private readonly settingsToggle = el<HTMLButtonElement>('settings-toggle');
  private readonly settingsPanel = el('settings-panel');
  private readonly gestureSelect = el<HTMLSelectElement>('gesture-mode');
  private readonly gestureNote = el('gesture-note');
  private readonly smoothingInput = el<HTMLInputElement>('smoothing');
  private readonly smoothingValue = el<HTMLOutputElement>('smoothing-value');
  private readonly cameraSelect = el<HTMLSelectElement>('camera-select');
  private readonly mirrorCheck = el<HTMLInputElement>('opt-mirror');
  private readonly skeletonCheck = el<HTMLInputElement>('opt-skeleton');
  private readonly videoCheck = el<HTMLInputElement>('opt-video');

  private readonly gate = el('gate');
  private readonly gateTitle = el('gate-title');
  private readonly gateBody = el('gate-body');
  private readonly gateAction = el<HTMLButtonElement>('gate-action');
  private readonly gateFine = el('gate-fine');

  private readonly toastNode = el('toast');
  private toastTimer = 0;

  private swatchButtons: HTMLButtonElement[] = [];
  private customSwatch!: HTMLLabelElement;
  private customInput!: HTMLInputElement;

  private gateState: GateState = 'idle';

  constructor(private readonly handlers: ControlHandlers) {
    this.settings = loadSettings();
    this.buildSwatches();
    this.bind();
    this.syncAll();
  }

  get activeWidth(): number {
    return this.settings.tool === 'eraser' ? this.settings.eraserWidth : this.settings.penWidth;
  }

  // ---- construction -----------------------------------------------------

  private buildSwatches(): void {
    for (const [index, color] of PALETTE.entries()) {
      const button = document.createElement('button');
      button.className = 'swatch';
      button.type = 'button';
      button.role = 'radio';
      button.style.background = color;
      button.dataset.color = color;
      button.title = `${color} — ${index + 1}`;
      button.setAttribute('aria-label', `Colour ${color}`);
      button.addEventListener('click', () => this.pickColor(color));
      this.swatchRow.append(button);
      this.swatchButtons.push(button);
    }

    const label = document.createElement('label');
    label.className = 'swatch swatch--custom';
    label.title = 'Custom colour';

    const input = document.createElement('input');
    input.type = 'color';
    input.value = this.settings.color;
    input.setAttribute('aria-label', 'Custom colour');
    input.addEventListener('input', () => this.pickColor(input.value));

    label.append(input);
    this.swatchRow.append(label);
    this.customSwatch = label;
    this.customInput = input;
  }

  private bind(): void {
    for (const button of this.toolButtons) {
      button.addEventListener('click', () => this.setTool(button.dataset.tool as Tool));
    }

    this.sizeInput.addEventListener('input', () => {
      this.setWidth(Number(this.sizeInput.value));
    });

    this.undoButton.addEventListener('click', () => this.handlers.onUndo());
    this.redoButton.addEventListener('click', () => this.handlers.onRedo());
    this.clearButton.addEventListener('click', () => this.handlers.onClear());
    this.saveButton.addEventListener('click', () => this.handlers.onSave('ink'));

    this.saveMenuToggle.addEventListener('click', (event) => {
      event.stopPropagation();
      this.toggleSaveMenu(this.saveMenu.hidden);
    });

    for (const item of this.saveMenu.querySelectorAll<HTMLButtonElement>('[data-export]')) {
      item.addEventListener('click', () => {
        this.toggleSaveMenu(false);
        this.handlers.onSave(item.dataset.export as ExportMode);
      });
    }

    this.settingsToggle.addEventListener('click', (event) => {
      event.stopPropagation();
      this.toggleSettings(this.settingsPanel.hidden);
    });

    document.addEventListener('click', (event) => {
      const target = event.target as Node;
      if (!this.saveMenu.hidden && !this.saveMenu.contains(target)) this.toggleSaveMenu(false);
      if (!this.settingsPanel.hidden && !this.settingsPanel.contains(target)) {
        this.toggleSettings(false);
      }
    });

    this.gestureSelect.addEventListener('change', () => {
      this.settings.gestureMode = this.gestureSelect.value as GestureMode;
      this.gestureNote.textContent = GESTURE_NOTES[this.settings.gestureMode];
      this.commit('gestureMode');
    });

    this.smoothingInput.addEventListener('input', () => {
      this.settings.smoothing = Number(this.smoothingInput.value) / 100;
      this.syncSmoothing();
      this.commit('smoothing');
    });

    this.cameraSelect.addEventListener('change', () => {
      this.settings.cameraId = this.cameraSelect.value;
      this.commit('cameraId');
    });

    this.mirrorCheck.addEventListener('change', () => {
      this.settings.mirror = this.mirrorCheck.checked;
      this.stage.classList.toggle('stage--mirrored', this.settings.mirror);
      this.commit('mirror');
    });

    this.skeletonCheck.addEventListener('change', () => {
      this.settings.showSkeleton = this.skeletonCheck.checked;
      this.commit('showSkeleton');
    });

    this.videoCheck.addEventListener('change', () => {
      this.settings.showVideo = this.videoCheck.checked;
      this.stage.classList.toggle('stage--novideo', !this.settings.showVideo);
      this.commit('showVideo');
    });

    this.gateAction.addEventListener('click', () => this.handlers.onStart());

    window.addEventListener('keydown', (event) => this.onKeyDown(event));
  }

  // ---- state ------------------------------------------------------------

  private commit(key: keyof Settings): void {
    saveSettings(this.settings);
    this.handlers.onSettingsChange(this.settings, key);
  }

  private setTool(tool: Tool): void {
    if (tool !== 'pen' && tool !== 'eraser') return;
    this.settings.tool = tool;
    this.syncTool();
    this.syncSize();
    this.commit('tool');
  }

  private pickColor(color: string): void {
    this.settings.color = color.toLowerCase();
    // Choosing a colour is also a statement of intent to draw, not erase.
    if (this.settings.tool === 'eraser') {
      this.settings.tool = 'pen';
      this.syncTool();
      this.syncSize();
    }
    this.syncColor();
    this.commit('color');
  }

  private setWidth(width: number): void {
    const clamped = Math.min(Math.max(Math.round(width), STROKE.minWidth), STROKE.maxWidth);
    if (this.settings.tool === 'eraser') this.settings.eraserWidth = clamped;
    else this.settings.penWidth = clamped;
    this.syncSize();
    this.commit(this.settings.tool === 'eraser' ? 'eraserWidth' : 'penWidth');
  }

  private nudgeWidth(delta: number): void {
    this.setWidth(this.activeWidth + delta);
  }

  // ---- syncing ----------------------------------------------------------

  private syncAll(): void {
    this.syncTool();
    this.syncColor();
    this.syncSize();
    this.syncSmoothing();

    this.gestureSelect.value = this.settings.gestureMode;
    this.gestureNote.textContent = GESTURE_NOTES[this.settings.gestureMode];
    this.mirrorCheck.checked = this.settings.mirror;
    this.skeletonCheck.checked = this.settings.showSkeleton;
    this.videoCheck.checked = this.settings.showVideo;
    this.stage.classList.toggle('stage--mirrored', this.settings.mirror);
    this.stage.classList.toggle('stage--novideo', !this.settings.showVideo);
  }

  private syncTool(): void {
    for (const button of this.toolButtons) {
      button.setAttribute('aria-checked', String(button.dataset.tool === this.settings.tool));
    }
  }

  private syncColor(): void {
    document.documentElement.style.setProperty('--ink', this.settings.color);
    this.customInput.value = this.settings.color;

    const current = this.settings.color.toLowerCase();
    let matched = false;
    for (const button of this.swatchButtons) {
      const isActive = button.dataset.color?.toLowerCase() === current;
      button.setAttribute('aria-checked', String(isActive));
      matched ||= isActive;
    }
    this.customSwatch.setAttribute('aria-checked', String(!matched));
    this.customSwatch.classList.toggle('is-active', !matched);
    if (!matched) this.customSwatch.style.background = this.settings.color;
    else this.customSwatch.style.removeProperty('background');

    // Keeps a light custom swatch from vanishing into the panel edge.
    this.customSwatch.style.borderColor =
      !matched && luminance(this.settings.color) > 0.7 ? 'rgb(0 0 0 / 0.35)' : '';
  }

  private syncSize(): void {
    this.sizeInput.value = String(this.activeWidth);
    this.sizeValue.value = String(this.activeWidth);
    paintRange(this.sizeInput);
  }

  private syncSmoothing(): void {
    this.smoothingInput.value = String(Math.round(this.settings.smoothing * 100));
    this.smoothingValue.value = `${Math.round(this.settings.smoothing * 100)}%`;
    paintRange(this.smoothingInput);
  }

  // ---- panels -----------------------------------------------------------

  private toggleSettings(open: boolean): void {
    this.settingsPanel.hidden = !open;
    this.settingsToggle.setAttribute('aria-expanded', String(open));
  }

  private toggleSaveMenu(open: boolean): void {
    this.saveMenu.hidden = !open;
    this.saveMenuToggle.setAttribute('aria-expanded', String(open));
  }

  // ---- public surface ---------------------------------------------------

  setHistory(stats: BoardStats): void {
    this.undoButton.disabled = !stats.canUndo;
    this.redoButton.disabled = !stats.canRedo;
    this.clearButton.disabled = stats.strokes === 0;
    this.saveButton.disabled = stats.strokes === 0;
    this.saveMenuToggle.disabled = stats.strokes === 0;
    if (stats.strokes === 0) this.toggleSaveMenu(false);
  }

  setCameras(devices: MediaDeviceInfo[], activeId: string): void {
    this.cameraSelect.replaceChildren();

    const auto = document.createElement('option');
    auto.value = '';
    auto.textContent = 'Default camera';
    this.cameraSelect.append(auto);

    for (const [index, device] of devices.entries()) {
      const option = document.createElement('option');
      option.value = device.deviceId;
      option.textContent = device.label || `Camera ${index + 1}`;
      this.cameraSelect.append(option);
    }

    const known = devices.some((d) => d.deviceId === activeId);
    this.cameraSelect.value = known ? activeId : '';
    this.settings.cameraId = this.cameraSelect.value;
  }

  setGate(state: GateState, content?: GateContent): void {
    this.gateState = state;
    this.gate.hidden = state === 'hidden';
    this.gate.classList.toggle('gate--error', state === 'error');
    this.stage.classList.toggle('stage--error', state === 'error');

    if (!content) return;
    this.gateTitle.textContent = content.title;
    this.gateBody.textContent = content.body;
    this.gateFine.textContent = content.fine ?? '';
    this.gateFine.hidden = !content.fine;
    this.gateAction.hidden = !content.action;
    this.gateAction.disabled = state === 'busy';
    if (content.action) this.gateAction.textContent = content.action;
  }

  toast(message: string): void {
    this.toastNode.textContent = message;
    this.toastNode.classList.add('toast--on');
    window.clearTimeout(this.toastTimer);
    this.toastTimer = window.setTimeout(() => {
      this.toastNode.classList.remove('toast--on');
    }, 2200);
  }

  // ---- keyboard ---------------------------------------------------------

  private onKeyDown(event: KeyboardEvent): void {
    const target = event.target as HTMLElement | null;
    const typing =
      target instanceof HTMLInputElement ||
      target instanceof HTMLSelectElement ||
      target instanceof HTMLTextAreaElement;

    if (event.key === 'Escape') {
      this.toggleSaveMenu(false);
      this.toggleSettings(false);
      return;
    }

    const accel = event.metaKey || event.ctrlKey;

    if (accel && event.key.toLowerCase() === 'z') {
      event.preventDefault();
      if (event.shiftKey) this.handlers.onRedo();
      else this.handlers.onUndo();
      return;
    }

    if (accel && event.key.toLowerCase() === 'y') {
      event.preventDefault();
      this.handlers.onRedo();
      return;
    }

    if (accel || event.altKey || typing) return;

    switch (event.key.toLowerCase()) {
      case 'p':
        this.setTool('pen');
        break;
      case 'e':
        this.setTool('eraser');
        break;
      case '[':
        this.nudgeWidth(-2);
        break;
      case ']':
        this.nudgeWidth(2);
        break;
      case 'c':
        this.handlers.onClear();
        break;
      case 's':
        this.handlers.onSave('ink');
        break;
      case 'h':
        document.body.classList.toggle('ui-hidden');
        break;
      case 'enter':
      case ' ':
        if (this.gateState === 'idle') {
          event.preventDefault();
          this.handlers.onStart();
        }
        return;
      default: {
        const index = Number(event.key) - 1;
        if (Number.isInteger(index) && index >= 0 && index < PALETTE.length) {
          this.pickColor(PALETTE[index]);
        }
        return;
      }
    }
  }
}

File: drawone/ui/controls.py
"""Keyboard handling, the dock, the settings readout, the gate and toasts.

The web build had a DOM for this. Here the same controls are painted onto the
frame and driven from the keyboard, which is why every state the dock shows also
has a key that changes it.
"""

from __future__ import annotations

import time
from dataclasses import dataclass
from typing import Callable, List, Optional, Sequence, Tuple

import numpy as np

from ..board.board import BoardStats
from ..config import PALETTE, STROKE
from ..render.color import luminance, to_bgr
from ..types import Tool
from .settings import MODES, Settings, load_settings, save_settings
from .widgets import (
    LABEL,
    Palette,
    Rect,
    chip,
    meter,
    outline,
    panel,
    rule,
    text,
    text_size,
)

GESTURE_NOTES = {
    "pinch": "Touch your thumb to your index fingertip to ink. Separate them to lift the pen.",
    "point": "Extend your index finger and fold the other three to ink.",
    "grip": "Pinch with your ring and little fingers tucked in. Hardest to trigger by accident.",
    "any": "Any of the three gestures inks. Convenient, but easier to start a stroke by mistake.",
}

#: Single source of truth for the keyboard: what the help panel lists is what
#: `handle_key` dispatches.
KEYMAP: Sequence[Tuple[str, str]] = (
    ("P / E", "pen / eraser"),
    ("[ / ]", "brush size"),
    ("1-8", "palette colour"),
    ("Z / Y", "undo / redo"),
    ("C", "clear (undoable)"),
    ("S", "save transparent PNG"),
    ("shift+S", "save PNG with camera frame"),
    ("G", "gesture mode"),
    (", / .", "smoothing"),
    ("M / K / V", "mirror / skeleton / video"),
    ("N", "next camera"),
    ("H", "hide the controls"),
    ("?", "this list"),
    ("Q / Esc", "quit"),
)

TOAST_MS = 2200.0


@dataclass
class ControlHandlers:
    on_settings_change: Callable[[Settings, str], None] = lambda s, k: None
    on_undo: Callable[[], None] = lambda: None
    on_redo: Callable[[], None] = lambda: None
    on_clear: Callable[[], None] = lambda: None
    on_save: Callable[[str], None] = lambda mode: None
    on_next_camera: Callable[[], None] = lambda: None
    on_quit: Callable[[], None] = lambda: None


@dataclass
class GateContent:
    title: str
    body: str = ""
    fine: str = ""
    action: str = ""


class Controls:
    def __init__(self, handlers: ControlHandlers) -> None:
        self._handlers = handlers
        self.settings = load_settings()
        self._stats = BoardStats(0, False, False)
        self._hidden = False
        self._help = False
        self._gate_state = "hidden"
        self._gate: Optional[GateContent] = None
        self._toast = ""
        self._toast_at = -1e9

    @property
    def active_width(self) -> int:
        return self.settings.active_width

    @property
    def gate_state(self) -> str:
        return self._gate_state

    # ---- state ------------------------------------------------------------

    def set_history(self, stats: BoardStats) -> None:
        self._stats = stats

    def set_gate(self, state: str, content: Optional[GateContent] = None) -> None:
        self._gate_state = state
        if content is not None:
            self._gate = content

    def toast(self, message: str) -> None:
        self._toast = message
        self._toast_at = time.perf_counter() * 1000.0

    def _commit(self, key: str) -> None:
        save_settings(self.settings)
        self._handlers.on_settings_change(self.settings, key)

    def _set_tool(self, tool: Tool) -> None:
        if self.settings.tool == tool:
            return
        self.settings.tool = tool
        self._commit("tool")

    def _pick_color(self, color: str) -> None:
        self.settings.color = color.lower()
        # Choosing a colour is also a statement of intent to draw, not erase.
        if self.settings.tool == "eraser":
            self.settings.tool = "pen"
        self._commit("color")

    def _set_width(self, width: float) -> None:
        clamped = int(min(max(round(width), STROKE.min_width), STROKE.max_width))
        if self.settings.tool == "eraser":
            self.settings.eraser_width = clamped
            self._commit("eraser_width")
        else:
            self.settings.pen_width = clamped
            self._commit("pen_width")

    def _nudge_width(self, delta: int) -> None:
        self._set_width(self.active_width + delta)

    def _cycle_gesture(self) -> None:
        index = MODES.index(self.settings.gesture_mode)
        mode = MODES[(index + 1) % len(MODES)]
        self.settings.gesture_mode = mode  # type: ignore[assignment]
        self._commit("gesture_mode")
        self.toast("{} - {}".format(mode, GESTURE_NOTES[mode]))

    def _nudge_smoothing(self, delta: float) -> None:
        self.settings.smoothing = min(max(self.settings.smoothing + delta, 0.0), 1.0)
        self._commit("smoothing")

    # ---- keyboard ---------------------------------------------------------

    def handle_key(self, key: int) -> bool:
        """Returns False when the app should quit."""
        if key < 0:
            return True
        code = key & 0xFF

        if code in (27, ord("q")):  # Esc
            if code == 27 and self._help:
                self._help = False
                return True
            self._handlers.on_quit()
            return False

        char = chr(code) if 32 <= code < 127 else ""

        if char in ("?", "/"):
            self._help = not self._help
            return True
        if char == "p":
            self._set_tool("pen")
        elif char == "e":
            self._set_tool("eraser")
        elif char == "[":
            self._nudge_width(-2)
        elif char == "]":
            self._nudge_width(2)
        elif char == "z":
            self._handlers.on_undo()
        elif char in ("Z", "y", "Y"):
            self._handlers.on_redo()
        elif char == "c":
            self._handlers.on_clear()
        elif char == "s":
            self._handlers.on_save("ink")
        elif char == "S":
            self._handlers.on_save("composite")
        elif char == "h":
            self._hidden = not self._hidden
        elif char == "g":
            self._cycle_gesture()
        elif char == ",":
            self._nudge_smoothing(-0.05)
        elif char == ".":
            self._nudge_smoothing(0.05)
        elif char == "m":
            self.settings.mirror = not self.settings.mirror
            self._commit("mirror")
        elif char == "k":
            self.settings.show_skeleton = not self.settings.show_skeleton
            self._commit("show_skeleton")
        elif char == "v":
            self.settings.show_video = not self.settings.show_video
            self._commit("show_video")
        elif char == "n":
            self._handlers.on_next_camera()
        elif char.isdigit():
            index = int(char) - 1
            if 0 <= index < len(PALETTE):
                self._pick_color(PALETTE[index])

        return True

    # ---- rendering --------------------------------------------------------

    def render(self, image: np.ndarray, scale: float) -> None:
        # The gate is modal: while it is up there is no session to control, so
        # nothing else competes with it for the frame.
        if self._gate_state != "hidden" and self._gate is not None:
            self._draw_gate(image, scale)
            self._draw_toast(image, scale)
            return

        if not self._hidden:
            self._draw_dock(image, scale)
            self._draw_settings(image, scale)
        if self._help:
            self._draw_help(image, scale)
        self._draw_toast(image, scale)

    def _draw_dock(self, image: np.ndarray, scale: float) -> None:
        height, width = image.shape[:2]

        def s(value: float) -> int:
            return int(round(value * scale))

        chip_w, chip_h = s(58), s(24)
        swatch = s(17)
        swatch_gap = s(5)
        gap = s(8)
        group_gap = s(16)
        meter_w = s(74)

        tools_w = chip_w * 2 + gap
        swatches_w = len(PALETTE) * swatch + (len(PALETTE) - 1) * swatch_gap
        size_w = s(30) + meter_w + s(28)
        actions_w = chip_w * 4 + gap * 3
        content = tools_w + swatches_w + size_w + actions_w + group_gap * 3
        pad = s(14)

        dock_w = content + pad * 2
        dock_h = chip_h + pad * 2
        x0 = (width - dock_w) // 2
        y0 = height - dock_h - s(34)
        panel(image, (x0, y0, x0 + dock_w, y0 + dock_h), Palette.case, 0.82,
              border=Palette.rule)

        x = x0 + pad
        row_y = y0 + pad
        label_size = 0.36 * scale

        chip(image, "PEN", (x, row_y, x + chip_w, row_y + chip_h),
             active=self.settings.tool == "pen",
             accent=to_bgr(self.settings.color), size=label_size)
        x += chip_w + gap
        chip(image, "ERASER", (x, row_y, x + chip_w, row_y + chip_h),
             active=self.settings.tool == "eraser", accent=Palette.paper,
             size=label_size)
        x += chip_w + group_gap

        for color in PALETTE:
            rect: Rect = (x, row_y + s(3), x + swatch, row_y + s(3) + swatch)
            panel(image, rect, to_bgr(color), 1.0)
            if color.lower() == self.settings.color.lower():
                # A light swatch would vanish into the panel without this.
                edge = Palette.void if luminance(color) > 0.7 else Palette.paper
                outline(image, (rect[0] - s(2), rect[1] - s(2), rect[2] + s(2),
                                rect[3] + s(2)), edge)
            x += swatch + swatch_gap
        x += group_gap - swatch_gap

        text(image, "SIZE", x, row_y + s(7), Palette.dim2, label_size, 1, LABEL)
        x += s(30)
        span = STROKE.max_width - STROKE.min_width
        meter(image, (x, row_y + s(9), x + meter_w, row_y + s(14)),
              (self.active_width - STROKE.min_width) / span,
              to_bgr(self.settings.color) if self.settings.tool == "pen" else Palette.paper)
        x += meter_w + s(6)
        text(image, str(self.active_width), x, row_y + s(7), Palette.paper,
             label_size, 1, LABEL)
        x += s(22) + group_gap

        actions = (
            ("UNDO", self._stats.can_undo),
            ("REDO", self._stats.can_redo),
            ("CLEAR", self._stats.strokes > 0),
            ("SAVE", self._stats.strokes > 0),
        )
        for label, enabled in actions:
            chip(image, label, (x, row_y, x + chip_w, row_y + chip_h),
                 enabled=enabled, size=label_size)
            x += chip_w + gap

        hint = "P pen  E eraser  [ ] size  1-8 colour  Z undo  C clear  S save  ? keys"
        text(image, hint, width / 2, y0 + dock_h + s(8), Palette.dim2,
             0.33 * scale, 1, LABEL, align="center")

    def _draw_settings(self, image: np.ndarray, scale: float) -> None:
        def s(value: float) -> int:
            return int(round(value * scale))

        rows = (
            ("GESTURE", self.settings.gesture_mode),
            ("SMOOTH", "{}%".format(round(self.settings.smoothing * 100))),
            ("MIRROR", "on" if self.settings.mirror else "off"),
            ("SKELETON", "on" if self.settings.show_skeleton else "off"),
            ("VIDEO", "on" if self.settings.show_video else "off"),
            ("CAMERA", str(self.settings.camera_id)),
        )

        width = image.shape[1]
        pad = s(12)
        box_w = s(132)
        line = s(15)
        box_h = line * len(rows) + s(14)
        x0 = width - box_w - pad
        y0 = pad
        panel(image, (x0, y0, x0 + box_w, y0 + box_h), Palette.case, 0.72,
              border=Palette.rule)

        size = 0.34 * scale
        y = y0 + s(7)
        for label, value in rows:
            text(image, label, x0 + s(10), y, Palette.dim2, size, 1, LABEL)
            text(image, value, x0 + box_w - s(10), y, Palette.dim, size, 1, LABEL,
                 align="right")
            y += line

    def _draw_help(self, image: np.ndarray, scale: float) -> None:
        def s(value: float) -> int:
            return int(round(value * scale))

        height, width = image.shape[:2]
        line = s(19)
        box_w = s(330)
        box_h = line * len(KEYMAP) + s(46)
        x0 = (width - box_w) // 2
        y0 = (height - box_h) // 2
        panel(image, (x0, y0, x0 + box_w, y0 + box_h), Palette.case, 0.94,
              border=Palette.rule2)

        size = 0.38 * scale
        text(image, "KEYS", x0 + s(18), y0 + s(14), Palette.paper, size * 1.1, 1, LABEL)
        rule(image, x0 + s(18), y0 + s(32), x0 + box_w - s(18))

        y = y0 + s(40)
        for combo, description in KEYMAP:
            text(image, combo, x0 + s(18), y, Palette.trace, size, 1, LABEL)
            text(image, description, x0 + s(128), y, Palette.dim, size, 1, LABEL)
            y += line

    def _draw_gate(self, image: np.ndarray, scale: float) -> None:
        assert self._gate is not None
        gate = self._gate

        def s(value: float) -> int:
            return int(round(value * scale))

        height, width = image.shape[:2]
        panel(image, (0, 0, width, height), Palette.void, 0.8)

        error = self._gate_state == "error"
        title_size = 0.66 * scale
        body_size = 0.4 * scale
        line = s(21)

        lines = _wrap(gate.body, body_size, int(width * 0.48))
        widths = [text_size(gate.title, title_size, 1, LABEL)[0]]
        widths += [text_size(entry, body_size, 1, LABEL)[0] for entry in lines]
        if gate.action:
            widths.append(text_size(gate.action, body_size, 1, LABEL)[0])
        if gate.fine:
            widths.append(text_size(gate.fine, body_size * 0.9, 1, LABEL)[0])

        pad = s(30)
        box_w = min(max(widths) + pad * 2, width - s(40))
        box_h = (
            s(34)
            + len(lines) * line
            + (line + s(8) if gate.action else 0)
            + (line if gate.fine else 0)
            + pad
        )
        x0 = (width - box_w) // 2
        y0 = (height - box_h) // 2
        panel(image, (x0, y0, x0 + box_w, y0 + box_h), Palette.case, 0.95,
              border=Palette.signal if error else Palette.rule2)

        centre = x0 + box_w / 2
        y = y0 + s(22)
        text(image, gate.title, centre, y, Palette.signal if error else Palette.paper,
             title_size, 1, LABEL, align="center")
        y += s(34)
        for entry in lines:
            text(image, entry, centre, y, Palette.dim, body_size, 1, LABEL, align="center")
            y += line
        if gate.action:
            y += s(8)
            text(image, gate.action, centre, y, Palette.trace, body_size, 1, LABEL,
                 align="center")
            y += line
        if gate.fine:
            text(image, gate.fine, centre, y, Palette.dim2, body_size * 0.9, 1, LABEL,
                 align="center")

    def _draw_toast(self, image: np.ndarray, scale: float) -> None:
        if not self._toast:
            return
        elapsed = time.perf_counter() * 1000.0 - self._toast_at
        if elapsed > TOAST_MS:
            self._toast = ""
            return

        def s(value: float) -> int:
            return int(round(value * scale))

        height, width = image.shape[:2]
        size = 0.38 * scale
        text_w, text_h = text_size(self._toast, size, 1, LABEL)
        pad = s(12)
        box_w = text_w + pad * 2
        box_h = text_h + pad
        x0 = (width - box_w) // 2
        y0 = height - box_h - s(88)
        panel(image, (x0, y0, x0 + box_w, y0 + box_h), Palette.case2, 0.92,
              border=Palette.rule2)
        text(image, self._toast, width / 2, y0 + pad / 2 - s(1), Palette.paper,
             size, 1, LABEL, align="center")


def _wrap(value: str, size: float, max_width: int) -> List[str]:
    if not value:
        return []
    words = value.split()
    lines: List[str] = []
    current = ""
    for word in words:
        candidate = "{} {}".format(current, word).strip()
        if current and text_size(candidate, size, 1, LABEL)[0] > max_width:
            lines.append(current)
            current = word
        else:
            current = candidate
    if current:
        lines.append(current)
    return lines
